segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1375 @@
+                // Validate preconditions.
+                if (null == segmentMetadata || !segmentMetadata.isActive()) {
+                    throw new StreamSegmentNotExistsException(streamSegmentName);
+                }
+
+                if (segmentMetadata.getOwnerEpoch() > this.epoch) {
+                    throw new StorageNotPrimaryException(streamSegmentName);
+                }
+
+                // seal if it is not already sealed.
+                if (!segmentMetadata.isSealed()) {
+                    segmentMetadata.setSealed(true);
+                    txn.update(segmentMetadata);
+                    txn.commit();
+                }
+
+                log.debug("seal - segment={}.", handle.getSegmentName());
+                LoggerHelpers.traceLeave(log, "seal", traceId, handle);
+                return null;
+            } catch (StorageMetadataVersionMismatchException ex) {
+                throw ex;
+            } catch (StorageMetadataWritesFencedOutException ex) {
+                throw new StorageNotPrimaryException(streamSegmentName, ex);
+            }
+        });
+    }
+
+    /**
+     * Concatenates two StreamSegments together. The Source StreamSegment will be appended as one atomic block at the end
+     * of the Target StreamSegment (but only if its length equals the given offset), after which the Source StreamSegment
+     * will cease to exist. Prior to this operation, the Source StreamSegment must be sealed.
+     *
+     * @param targetHandle  A read-write SegmentHandle that points to the Target StreamSegment. After this operation
+     *                      is complete, this is the surviving StreamSegment.
+     * @param offset        The offset in the Target StreamSegment to concat at.
+     * @param sourceSegment The Source StreamSegment. This StreamSegment will be concatenated to the Target StreamSegment.
+     *                      After this operation is complete, this StreamSegment will no longer exist.
+     * @param timeout       Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate the operation succeeded. If the operation failed,
+     * it will contain the cause of the failure. Notable exceptions:
+     * <ul>
+     * <li> BadOffsetException: When the given offset does not match the actual length of the target segment in storage.
+     * <li> StreamSegmentNotExistsException: When the either the source Segment or the target Segment do not exist in Storage.
+     * <li> StorageNotPrimaryException: When this Storage instance is no longer primary for the target Segment (it was fenced out).
+     * <li> IllegalStateException: When the Source Segment is not Sealed.
+     * </ul>
+     * @throws IllegalArgumentException If targetHandle is read-only.
+     */
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration timeout) {
+        checkInitialized();
+        return execute( () -> {
```

Comment:
Method too long.

Reply:
Java doesn't have out parameters . So extracting smaller methods requires creating method objects etc. So instead of making code simple it makes it unnecessarily complicated and in my opinion defeats the purpose. If there is a simple way to break this method I'll definitely do that.